src/components/views/dialogs/ElementProDialog.tsx

```diff
@@ -0,0 +1,157 @@
+*/
+
+import * as React from "react";
+import BaseDialog from '../../views/dialogs/BaseDialog';
+import ElementProConfirmCloseDialog from './ElementProConfirmCloseDialog';
+import ElementProDataConfirmDialog from './ElementProDataConfirmDialog';
+import Modal from "../../../Modal";
+import SdkConfig from "../../../SdkConfig";
+import {_t} from "../../../languageHandler";
+import {MatrixClientPeg} from "../../../MatrixClientPeg";
+import {OwnProfileStore} from "../../../stores/OwnProfileStore";
+import {IPostmessage, IPostmessageResponseData, PostmessageAction} from "./ElementProDialogTypes";
+
+interface IProps {
+    requestClose(): void;
+}
+
+interface IState {
+    completed: boolean;
+    error: string;
+}
+
+export default class ElementProDialog extends React.PureComponent<IProps, IState> {
+    private iframeRef: React.RefObject<HTMLIFrameElement> = React.createRef();
+    private readonly elementProSetupUrl: string;
+
+    constructor(props: IProps) {
+        super(props);
+
+        this.state = {
+            completed: false,
+            error: null,
+        };
+
+        this.elementProSetupUrl = SdkConfig.get().element_pro.url;
+    }
+
+    private messageHandler = (message: IPostmessage) => {
+        if (!this.elementProSetupUrl.startsWith(message.origin)) {
+            return;
+        }
+        switch (message.data.action) {
+            case PostmessageAction.ElementProAccountDetailsRequest:
+                Modal.createDialog(
+                    ElementProDataConfirmDialog,
+                    {
+                        onFinished: result => {
+                            if (result) {
+                                return this.sendAccountDetails();
+                            }
+                        },
+                    },
+                );
+                break;
+            case PostmessageAction.SetupComplete:
+                // Set as completed but let the user close the modal themselves
+                // so they have time to finish reading any information
+                this.setState({
+                    completed: true,
+                });
+                break;
+            case PostmessageAction.CloseDialog:
+                this.onFinished(true);
+                break;
+        }
+    }
+
+    private onFinished = (result: boolean) => {
+        if (result || this.state.completed) {
+            // We're done, close
+            this.props.requestClose();
+        } else {
+            Modal.createDialog(
+                ElementProConfirmCloseDialog,
+                {
+                    onFinished: result => {
+                        if (result) {
+                            this.props.requestClose();
+                        }
+                    },
+                },
+            );
+        }
+    }
+
+    private sendMessage = (message: IPostmessageResponseData) => {
+        this.iframeRef.current.contentWindow.postMessage(message, this.elementProSetupUrl);
```

Comment:
I thought we were going to use the widget API in the end?

Reply:
We decided to not use it because it would mean any widget would be able to request an access token and users might accidentally give such permission. We decided to use a custom postmessage API for this feature since it has very specific purpose not necessary for widgets. This feature also relies on as little amount of prompts as possible to make the host setup experience nicer for those who choose to use this feature. Using the widget API separately to query openid token and separately for the Element Pro specific cases would introduce two dialogs. We would like to ask the user once for all that is needed for this feature without making permanent changes to the widget API.